File: services/csmoney.go
package services

import (
	"encoding/json"
	"fmt"
	"math/rand"
	"strconv"
	"time"

	"github.com/cyberbebebe/cs2-profit-checker/types"
	"github.com/cyberbebebe/cs2-profit-checker/utils"
	"github.com/enetx/g"
	shttp "github.com/enetx/http"
	"github.com/enetx/surf"
)

type CSMoneyService struct {
	Cookies []*shttp.Cookie
	Client  *surf.Client
}


func NewCSMoneyService(cookies map[string]string) *CSMoneyService {
	var cookieSlice []*shttp.Cookie
	for name, value := range cookies {
		cookieSlice = append(cookieSlice, &shttp.Cookie{
			Name:  name,
			Value: value,
		})
	}

	// Create Surf Client
	surfClient := surf.NewClient().
		Builder().
		Impersonate().Chrome().
		Session().
		Build().
		Unwrap()

	return &CSMoneyService{
		Cookies: cookieSlice,
		Client:  surfClient,
	}
}



func (s *CSMoneyService) Name() string {
	return "CSMoney"
}

func (s *CSMoneyService) GetSales(startTime, endTime time.Time) ([]types.Transaction, error){
	return s.fetchHistory("sold", startTime, endTime)
}

func (s *CSMoneyService) GetBuys(startTime, endTime time.Time) ([]types.Transaction, error){
	return s.fetchHistory("accepted", startTime, endTime)
}

func (s *CSMoneyService) fetchHistory(status string, startTime, endTime time.Time) ([]types.Transaction, error) {
	var allTxs []types.Transaction

	offset := ""
	acceptedStr := ""
	if status == "accepted"{
		acceptedStr = "&type=buy"
	}

	for{
		url := fmt.Sprintf("https://cs.money/2.0/market/history?limit=100&noCache=true%s&status=%s%s", offset, status, acceptedStr)
		g_url := g.String(url)
		
		req := s.Client.Get(g_url)
		for _, cookie := range s.Cookies {
			req = req.AddCookies(cookie)
		}
		resp := req.Do()

		if !resp.IsOk() {
			return allTxs, fmt.Errorf("request failed: %v", resp.Err())
		}

		response := resp.Ok()
		defer response.Body.Close()

		if response.StatusCode != 200 {
			return allTxs, fmt.Errorf("CSMoney error %d", response.StatusCode)
		}
		
		// Get body as bytes
		bodyResult := response.Body.Bytes()
		if !bodyResult.IsOk() {
			return allTxs, fmt.Errorf("failed to read body: %v", bodyResult.Err())
		}
		
		bodyBytes := bodyResult.Ok()
		
		var items csmoneyResponse
		if err := json.Unmarshal(bodyBytes.Std(), &items); err != nil {
			return allTxs, fmt.Errorf("decode error: %v", err)
		}

		for _, raw := range items {
			txs := s.convertCSMoneyTx(raw)
			for _, tx := range txs {
				if tx.Date.After(endTime) { continue }
				if tx.Date.Before(startTime) { return allTxs, nil }
				allTxs = append(allTxs, tx)
			}
		}
		
		lastItem := items[len(items)-1]
		newOffset := lastItem.Offset

		if len(items) < 100 {
			break
		}

		offset = fmt.Sprintf("&offset=%d", newOffset)

		sleepMs := 3000 + rand.Intn(1500) 
		time.Sleep(time.Duration(sleepMs) * time.Millisecond)
	}

	return allTxs, nil

}

func (s *CSMoneyService) convertCSMoneyTx(raw csmoneyItem) []types.Transaction {
	var transactions []types.Transaction
	
	date := time.Unix(raw.UpdateTime/1000, 0)
	if raw.TimeSettlement != nil {
        date = time.Unix(*raw.TimeSettlement, 0)
    }

	// SALES
	if raw.Type == "sell" {
		data := raw.Details.SellOrder.Skins
		
		finalPattern := -1
		if data.Asset.PaintSeed != nil{
			finalPattern = *data.Asset.PaintSeed
		}
		if data.Asset.CharmPattern != nil {
			finalPattern = *data.Asset.CharmPattern
		}

		fVal, _ := strconv.ParseFloat(data.Asset.Float, 64)
		cleanName, phase := utils.ExtractPhase(data.Asset.Names.Full)
		tx := types.Transaction{
			Source:    "CSMoney",
			Type:      types.TxSell,
			TxID:      raw.SourceID,
			AssetID:   data.Asset.ID,     
			ItemName:  cleanName,
			Price:     raw.Details.OnWallet,
			Currency:  "USD",
			Date:      date,
			FloatVal:  fVal,
			Pattern:   finalPattern,
			Phase:     phase,
		}
		// Signature
		tx.Signature = utils.GenerateSignature(tx.ItemName, tx.FloatVal, finalPattern)
		transactions = append(transactions, tx)

	// BUYS
	} else if raw.Type == "buy" {
		for _, skin := range raw.Details.Offer.Skins {
			cleanName, phase := utils.ExtractPhase(skin.Asset.Names.Full)

			finalPattern := -1
			if skin.Asset.Pattern != nil{
				finalPattern = *skin.Asset.Pattern
			}
			if skin.Asset.CharmPattern != nil {
				finalPattern = *skin.Asset.CharmPattern
		}

			tx := types.Transaction{
				Source:    "CSMoney",
				Type:      types.TxBuy,
				TxID:      fmt.Sprintf("%d", skin.ID),
				AssetID:   fmt.Sprintf("%d", skin.Asset.ID),
				ItemName:  cleanName,
				Price:     skin.Pricing.Computed,
				Currency:  "USD",
				Date:      date,
				FloatVal:  skin.Asset.Float,
				Pattern:   finalPattern,
				Phase:     phase,
			}
			// Signature
			tx.Signature = utils.GenerateSignature(tx.ItemName, tx.FloatVal, finalPattern)
			transactions = append(transactions, tx)
		}
	}

	return transactions
}

File: js/fetchers/csmoney.js
import { BaseFetcher, Transaction } from "./base.js";

// Helper for extracting Doppler Phases from item name
function extractPhase(name) {
  if (!name.includes("Doppler")) return { name, phase: "" };

  let phase = "";
  if (name.includes("Phase 1")) phase = "Phase 1";
  else if (name.includes("Phase 2")) phase = "Phase 2";
  else if (name.includes("Phase 3")) phase = "Phase 3";
  else if (name.includes("Phase 4")) phase = "Phase 4";
  else if (name.includes("Ruby")) phase = "Ruby";
  else if (name.includes("Sapphire")) phase = "Sapphire";
  else if (name.includes("Black Pearl")) phase = "Black Pearl";
  else if (name.includes("Emerald")) phase = "Emerald";

  if (phase) {
    // Remove phase from name (e.g. "Karambit | Doppler (Factory New) Phase 1" -> "Karambit | Doppler (Factory New)")
    // Usually phase is at the end or in brackets, simple replace might leave double spaces
    const cleanName = name.replace(phase, "").trim();
    return { name: cleanName, phase };
  }
  return { name, phase: "" };
}

export class CSMoneyFetcher extends BaseFetcher {
  constructor() {
    super("CSMoney");
  }

  async checkSession() {
    try {
      await this.fetchWithAuth("https://cs.money/1.0/market/user-store");
      return true;
    } catch (e) {
      return false;
    }
  }

  async getSales() {
    const sold = await this.fetchHistory("sold", "sell");
    const protectedSales = await this.fetchHistory("trade_protected", "sell");
    return [...sold, ...protectedSales];
  }

  async getBuys() {
    const buy = await this.fetchHistory("accepted", "buy");
    const protectedBuys = await this.fetchHistory("trade_protected", "buy");
    return [...buy, ...protectedBuys];
  }

  async fetchHistory(status, type) {
    let allTxs = [];
    let offsetParam = "";

    let typeParam = "";
    if (type === "buy") typeParam = "&type=buy";
    if (type === "sell") typeParam = "&type=sell"; // API might default to sell, but explicit is better

    while (true) {
      const url = `https://cs.money/2.0/market/history?limit=100&noCache=true${offsetParam}&status=${status}${typeParam}`;

      try {
        const data = await this.fetchWithAuth(url);

        if (!Array.isArray(data)) {
          if (data.error) throw new Error(data.error);
          break;
        }

        if (data.length === 0) break;

        for (const raw of data) {
          const parsedTxs = this.convertCSMoneyTx(raw);
          allTxs.push(...parsedTxs);
        }

        // Pagination
        if (data.length < 100) break;

        const lastItem = data[data.length - 1];
        if (lastItem && lastItem.offset) {
          offsetParam = `&offset=${lastItem.offset-1}`;
        } else {
          break;
        }

        await this.sleep(1500);
      } catch (e) {
        console.error(`[CSMoney] Error fetching ${status}/${type}:`, e);
        break;
      }
    }

    return allTxs;
  }

  convertCSMoneyTx(raw) {
    const transactions = [];

    // Dates
    // UpdateTime is ms
    let createdDate = new Date(raw.updateTime);
    let verifiedDate = null;

    if (raw.timeSettlement) {
      verifiedDate = new Date(raw.timeSettlement * 1000);
    }

    // 1. SELL Logic
    if (raw.type === "sell" && raw.details && raw.details.sellOrder) {
      const skins = raw.details.sellOrder.skins;
      const asset = skins.asset;

      let finalPattern = -1;
      if (asset.paintseed !== undefined && asset.paintseed !== null) {
        finalPattern = asset.paintseed;
      }
      if (
        asset.keychainPattern !== undefined &&
        asset.keychainPattern !== null
      ) {
        finalPattern = asset.keychainPattern;
      }

      const floatVal = parseFloat(asset.float || 0);
      const { name: cleanName, phase } = extractPhase(asset.names.full);

      // Price on wallet (income)
      const price = raw.details.onWallet;

      transactions.push(
        new Transaction({
          source: "CSMoney",
          type: "SELL",
          tx_id: raw.sourceId, // Using sourceId as TxID
          asset_id: asset.id,
          item_name: cleanName,
          price: price,
          currency: "USD",
          created_at: createdDate,
          verified_at: verifiedDate, // Might be null if trade_protected
          float_val: floatVal,
          pattern: finalPattern,
          phase: phase,
        }),
      );
    }

    // 2. BUY Logic
    else if (raw.type === "buy" && raw.details && raw.details.offer) {
      const offerSkins = raw.details.offer.skins || [];

      for (const skin of offerSkins) {
        const asset = skin.asset;
        const { name: cleanName, phase } = extractPhase(asset.names.full);

        let finalPattern = -1;
        if (asset.pattern !== undefined && asset.pattern !== null) {
          finalPattern = asset.pattern;
        }
        if (
          asset.keychainPattern !== undefined &&
          asset.keychainPattern !== null
        ) {
          finalPattern = asset.keychainPattern;
        }

        const price = skin.pricing ? skin.pricing.computed : 0;

        transactions.push(
          new Transaction({
            source: "CSMoney",
            type: "BUY",
            tx_id: skin.id.toString(), // Individual Skin ID as TxID
            asset_id: asset.id.toString(),
            item_name: cleanName,
            price: price,
            currency: "USD",
            created_at: createdDate,
            verified_at: verifiedDate || createdDate,
            float_val: asset.float,
            pattern: finalPattern,
            phase: phase,
          }),
        );
      }
    }

    return transactions;
  }
}
